src/inference/dev_api/openvino/runtime/plugin_config.hpp

```diff
@@ -0,0 +1,416 @@
+                          << std::endl;
+                option.set_any(val);
+            }
+        }
+
+        OptionType& m_option;
+    };
+
+    virtual void apply_model_specific_options(const IRemoteContext* context, const ov::Model& model) {}
+    void apply_env_options();
+    void apply_config_options(std::string_view device_name, std::string_view config_path = "");
+    virtual void finalize_impl(const IRemoteContext* context) {}
+
+    template <typename T, PropertyMutability mutability>
+    bool is_set_by_user(const ov::Property<T, mutability>& property) const {
+        return m_user_properties.find(property.name()) != m_user_properties.end();
+    }
+
+    ConfigOptionBase* get_option_ptr(const std::string& name) const {
+        auto it = m_options_map.find(name);
+        OPENVINO_ASSERT(it != m_options_map.end(), "Option not found: ", name);
+        OPENVINO_ASSERT(it->second != nullptr, "Option is invalid: ", name);
+
+        return it->second;
+    }
+
+    template <typename T, PropertyMutability mutability>
+    void apply_rt_info_property(const ov::Property<T, mutability>& property, const ov::RTMap& rt_info) {
+        if (!is_set_by_user(property)) {
+            auto rt_info_val = rt_info.find(property.name());
+            if (rt_info_val != rt_info.end()) {
+                set_user_property({property(rt_info_val->second.template as<T>())}, OptionVisibility::RELEASE);
+            }
+        }
+    }
+
+    ov::AnyMap read_config_file(std::string_view filename, std::string_view target_device_name) const;
+    ov::AnyMap read_env() const;
+    static ov::Any read_env(const std::string& option_name, std::string_view prefix, const ConfigOptionBase* option);
+    void cleanup_unsupported(ov::AnyMap& config) const;
+
+    std::map<std::string, ConfigOptionBase*> m_options_map;
+    using OptionMapEntry = decltype(m_options_map)::value_type;
+
+    // List of properties explicitly set by user via Core::set_property() or Core::compile_model() or ov::Model's
+    // runtime info
+    ov::AnyMap m_user_properties;
+
+    std::string_view get_help_message(const std::string& name = "") const;
+    void print_help() const;
+
+    bool m_is_finalized = false;
+
+    inline static const std::string_view m_allowed_env_prefix = "OV_";
+};
+
+class OptionRegistrationHelper {
+public:
+    OptionRegistrationHelper(PluginConfig* config, std::string_view name, ConfigOptionBase* option) {
+        if (config)
+            config->register_option(std::string{name}, option);
+    }
+};
+
+template <typename T>
+struct TypedOption : public ConfigOptionBase {
+    TypedOption(const T& default_val, std::string_view prop_name, std::string_view desc)
+        : ConfigOptionBase(prop_name, desc),
+          value(default_val) {}
+    T value;
+};
+
+template <typename T, OptionVisibility visibility_>
+struct ConfigOption : public TypedOption<T> {
```

Comment:
Maybe have two classes with and without validation (less checks inside)
The class with validation can inherit or use without but applied validation.
?

Reply:
There's a little code which checks validator. Having sub-class will introduce much more, especially for options registration part where current impls just allow forwarding arguments from macro to option c-tor